### PR TITLE
feat: Add routing replay support for FSDP

### PR DESCRIPTION
**Pull request by XianlongLi**
# PR type
- [ ] Bug Fix
- [x] New Feature
- [ ] Document Updates
- [ ] More Models or Datasets Support

# PR information

This PR introduces routing replay support for Mixture-of-Experts (MoE) models when using FSDP (Fully Sharded Data Parallel) backend in the twinkle framework. It enables three replay modes (disabled, R2, and R3) to ensure consistent expert routing during reinforcement learning training. The feature is demonstrated through a GRPO training example on math tasks.

**Key Changes:**

**Core Router Replay Infrastructure** (`src/twinkle/model/transformers/moe/router_replay.py`):
   - Introduces `RouterReplayAction` enum defining three replay modes: RECORD, REPLAY_FORWARD, and REPLAY_BACKWARD
   - Implements a global registry system for per-MoE-block replay state management
   - Provides APIs to record routing decisions, inject pre-computed routings, and retrieve routing data

 **Expert Parallel Module Enhancements** (`src/twinkle/model/transformers/moe/expert_parallel.py`):
   - Adds block-level replay state integration into the router logic
   - Implements recording and replay of expert routing indices

 **TransformersModel Integration** (`src/twinkle/model/transformers/transformers.py`):
   - Adds `enable_router_replay` configuration option
   - Implements `_router_replay_setup()` method for managing replay lifecycle
   - Integrates replay into `forward()`, `forward_only()`, and `backward()` methods
   - Supports manual cleanup for `forward_backward()` operations

**Data Format Updates**:
   - Extends `InputFeature` with `routed_experts` field (`src/twinkle/data_format/input_feature.py`)
   - Adds `routed_experts` output support in `ModelOutput` (`src/twinkle/data_format/output.py`)
   - Adds `routed_experts` field to `SampledSequence` (`src/twinkle/data_format/sampling.py`)

**Input Processing Pipeline** (`src/twinkle/processor/base.py` ):
   - Adds `align_routed_experts()` preprocessing step to align vLLM routing data with input sequence lengths
   - Handles padding mismatch between routed_experts and input sequences (±1 token tolerance)
   - Updates padding and concatenation logic for the new field

**vLLM Sampler Integration** (`src/twinkle/sampler/vllm_sampler/` files):
   - Extracts and propagates `routed_experts` data from vLLM sampling output
   - Ensures routing data flows through the sampling pipeline

**Usage Example** (new files):
   - Complete GRPO training cookbook example with routing replay for GSM8K math problems (`cookbook/rl/grpo/short_math_grpo_routing_replay.py` - 334 lines)
   - Accompanying shell script for execution (`short_math_grpo_routing_replay.sh`)

**Replay Modes:**
- **R2**: Records expert routing during a forward-only RECORD pass, then replays during training
- **R3**: Uses routed_experts data returned by vLLM (v0.14.0+) for direct replay without recording
- **Disabled**: Standard behavior without routing replay

## Usage Tutorial
Refer to cookbook/rl/grpo/short_math_grpo_routing_replay.py

## Experiment results
kl
is1
is2

### CODE REVIEW

**Comment by tpx818** [comment]
/gemini review

Please focus on whether the routing returned in R3 mode and R2 mode is correct, and whether R3 mode correctly uses the vLLM expert routing.

File: src/twinkle/processor/base.py [hunk 1]
```diff
@@ -716,7 +720,8 @@ def is_mm_position_ids(position_ids):
                     num_axes = values[0].shape[0]
                     result[key] = result[key].reshape(len(values), num_axes, -1).permute(1, 0, 2).contiguous()
                 elif isinstance(values[0], torch.Tensor):
-                    result[key] = InputProcessor._pad_sequence(values, self.padding_map[key], self.padding_side)
+                    concat = False if (key == 'routed_experts') else None
+                    result[key] = InputProcessor._pad_sequence(values, self.padding_map[key], self.padding_side, concat)
```
[the remark this answers is not in the document]
Comment (XianlongLi) [reply src/twinkle/processor/base.py:724]: Not a valid question, will not be accepted.

File: src/twinkle/processor/base.py [hunk 2]
```diff
@@ -799,3 +804,28 @@ def postprocess_tensor_cp(self, tensor, cu_seqlens=None):
...
+                # The number of experts in the output can be 1 less than (prompt_length + response_token_count)
+                # This gap of 1 is expected
+                # For more details, please refer PR https://github.com/vllm-project/vllm/pull/28284
+                experts_seq_len = routed_experts.shape[0]
+                # Padding routed_experts(seq_len, layer_num, topk) seq_len to match the seq_len of the input_ids
+                padding_routed_experts = routed_experts
+                padding_len = input_seq_len - experts_seq_len
+                if padding_len > 0:
+                    padding_routed_experts = torch.nn.functional.pad(routed_experts, (0, 0, 0, 0, 0, padding_len),
+                                                                     'constant', self.padding_map.get('routed_experts', 0))
```
[the remark this answers is not in the document]
Comment (XianlongLi) [reply src/twinkle/processor/base.py]: Not a valid question, will not be accepted.

**Comment by XianlongLi** [comment]
@tpx818  Please review again.

**Review by tpx818 — APPROVED** [review]
